File: GVO Code v3/TimerUnit.pas
unit TimerUnit;

interface
uses SysUtils, Utils, windows;

Procedure TimerUpdate();
Procedure AddTicks();
procedure Padle_Timer_Update();

implementation
Uses main, slew, CalcCoord, HandPadle, Sync, OMS68SERMC;

procedure TimerUpdate();
Var
	CmdStr, CmdStr2, TempStr: String;
	Label EndTimer;
Begin {Timer Update}
	IF (NoPassword) then
	begin
		cmdstr := 'AA ST;';
		SendString(CommRecord, pchar(CmdStr));
		MainForm.Timer_Main.enabled := True;
		exit;
	end;

	MainForm.CmdEdit.Enabled := False;

	// Timer Off
	MainForm.Timer_Main.enabled := False;

	// ***********************if moving check for end of move***********************
	If (MovingRA) Then
	Begin
		SendAndGetString(CommRecord, 'AX QA;', Response);
		Tempstr := Response;
		if (length(tempstr) <> 4) Then exit;
		If (UpCase(Tempstr[2]) = 'D') Then
		Begin
			MovingRA := False;
			MainForm.cmdbx.Items.Insert(0, 'RA Axis Move Complete');
			Str(TrkRate:10:6, CmdStr2);
			CmdStr := 'AX JF' + CmdStr2 + ';';     //Back to Tracking
			SendString(CommRecord, pchar(CmdStr)); {RA Back To TRACKING}
			Coord(pndomem^).RAGoTo := 0.0;
			mainform.ejog.enabled := true;
			mainform.wjog.enabled := true;
		End;
	End;

	If (MovingDEC) Then
	Begin
		SendAndGetString(CommRecord, 'AY QA;', Response);
		Tempstr := Response;
		if (length(tempstr) <> 4) Then exit;
		If (UpCase(Tempstr[2]) = 'D') Then
		Begin
			MovingDEC := False;
			MainForm.cmdbx.Items.Insert(0, 'DEC Axis Move Complete');
			Coord(pndomem^).decGoTo := 0.0;
			mainform.sjog.enabled := true;
			mainform.njog.enabled := true;
		End;
	End;

	// *********************Update Coordinates*********************
	HalfSecondCounter := HalfSecondCounter + 1;
	If (HalfSecondCounter = 5) Then HalfSecondCounter := 0;
	If (HalfSecondCounter = 4) then UpdateCoord; {coordinates updated - at .5 sec}

	// ************************Leave if in motion**********************************
	If ((MovingRA) OR (MovingDec)) Then GoTo EndTimer;

	// ***************************Sync Scope****************************
	If ((Coord(pndomem^).RASync <> 0.0) AND (Coord(pndomem^).decSync <> 0.0)) Then
		SyncScope;  // sync requested

	// ********************** Slew if commanded ***********************
	If ((Coord(pndomem^).RAGoTo <> 0.0) AND (Coord(pndomem^).decGoTo <> 0.0)) Then
	Begin
		SlewScope;
		GoTO EndTimer;
	End;

	// *************Check for Park************
	If (parkit = True) Then
	Begin
		Parkit := False;
		mainform.sjog.enabled := False;
		mainform.njog.enabled := False;
		mainform.ejog.enabled := False;
		mainform.wjog.enabled := False;
		MovingRa := True;
		MovingDec := True;

		CmdStr := 'AX  kl; ';      {Send RA stop}
		SendString(CommRecord, pchar(CmdStr));
		sleep(300);
		CmdStr := 'AA vl75000,50000; MA0,0,,; GD; ID; ';      {Send home}
		SendString(CommRecord, pchar(CmdStr));
		MainForm.ButtonStop.Caption := 'START';
		GoTO EndTimer;
	End;

	// *****************handle padle***********
	HandleHandPadle;

	// *******************{Timer Back On}**********************
EndTimer:
	if (MainForm.ButtonStop.Caption = 'STOP') then
	begin
		MainForm.Timer_Main.enabled := True;
	end;
End; {Timer Update}

Procedure AddTicks();
Begin
	// Update Quadrant display
	MainForm.Quad1.Visible := False;
	MainForm.Quad2.Visible := False;
	MainForm.Quad3.Visible := False;
	MainForm.Quad4.Visible := False;

	If (Ypole > 0) Then MainForm.YPolarity.Caption := '+'
	Else MainForm.YPolarity.Caption := '-';

	If (Quadrant = 1) Then MainForm.Quad1.Visible := True
	Else If (Quadrant = 2) Then MainForm.Quad2.Visible := True
	Else If (Quadrant = 3) Then MainForm.Quad3.Visible := True
	Else If (Quadrant = 4) Then MainForm.Quad4.Visible := True;

	MainForm.Refresh;
End; {Add Ticks}

procedure Padle_Timer_Update();
Begin {Timer Update}
	// Timer Off
	MainForm.Timer_padle.Enabled := False;
	HandleHandPadle;
	MainForm.Timer_padle.Enabled := True;
End;

End.

File: GVO Code C++ v1/TimerUnit.cpp
#include "TimerUnit.h"

#include <cctype>
#include <cstdio>
#include <iostream>

#include "CalcCoord.h"
#include "CommUtils.h"
#include "GlobalValues.h"
#include "HandPaddle.h"
#include "OMS68SERMC.h"
#include "Slew.h"
#include "Sync.h"

extern char Response[256];

void TimerUpdate() {
	static bool inTimerUpdate = false;
	if (inTimerUpdate) {
		return;
	}
	inTimerUpdate = true;

	struct TimerGuard {
		bool& flag;
		~TimerGuard() { flag = false; }
	} guard{inTimerUpdate};

	Coord* CoordMem = CommUtils::GetCoordPtr();
	if (!CoordMem) {
		std::cout << "CoordMem is null\n";
		return;
	}

	// -------------------- If moving, check for end of move --------------------
	if (movingRA) {
		char axResp[256] = {};
		if (!SendAndGetCommand(&CommRecord, "AX QA;", axResp, sizeof(axResp))) {
			return;
		}

		std::string tempStr = axResp;

		if (tempStr.length() != 4) {
			return;
		}

		if (std::toupper(static_cast<unsigned char>(tempStr[1])) == 'D') {
			movingRA = false;
			std::cout << "RA Axis Move Complete\n";

			CoordMem->RAGoto = 0.0;

			if (!parkInProgress) {
				char buf[64];
				std::sprintf(buf, "%.6f", TrkRate);

				std::string cmdStr = "AX JF" + std::string(buf) + ";";
				std::cout << "[TimerUpdate] SEND: [" << cmdStr << "]\n";
				SendCommand(cmdStr);

				Xstate = StateVar::Tracking;
				Ystate = StateVar::Tracking;
				std::cout << "[TimerUpdate] Xstate/Ystate set to Tracking after move complete\n";
			}
			else {
				std::cout << "[TimerUpdate] Park move RA complete; not restarting tracking\n";
			}
		}
	}

	if (movingDEC) {
		char ayResp[256] = {};
		if (!SendAndGetCommand(&CommRecord, "AY QA;", ayResp, sizeof(ayResp))) {
			return;
		}

		std::string tempStr = ayResp;

		if (tempStr.length() != 4) {
			return;
		}

		if (std::toupper(static_cast<unsigned char>(tempStr[1])) == 'D') {
			movingDEC = false;
			std::cout << "DEC Axis Move Complete\n";

			CoordMem->DecGoto = 0.0;
		}
	}

	HalfSecondCounter++;
	if (HalfSecondCounter >= 5) {
		HalfSecondCounter = 0;
	}

	if (HalfSecondCounter == 4) {
		UpdateCoord();
	}

	if (movingRA || movingDEC) {
		return;
	}

	if (CoordMem->RASync != 0.0 && CoordMem->DecSync != 0.0) {
		SyncScope();
		return;
	}

	if (CoordMem->RAGoto != 0.0 && CoordMem->DecGoto != 0.0) {
		double raTarget = CoordMem->RAGoto;
		double decTarget = CoordMem->DecGoto;

		std::cout << "[TimerUpdate] starting slew request "
				  << "RAGoto=" << raTarget
				  << " DecGoto=" << decTarget
				  << "\n";

		SlewScope(raTarget, decTarget);
		return;
	}

	if (Parkit) {
		Parkit = false;

		movingRA = true;
		movingDEC = true;

		SendCommand("AX KL;");
		Sleep(300);
		std::string parkCmd =
			"AX AC" + xac + "; "
			"AY AC" + yac + "; "
			"AA VL" + xvlslew + "," + yvlslew + "; "
			"MA0,0,,; GD; ID;";
			
		SendCommand("AA VL75000,50000; MA0,0,,; GD; ID;");

		return;
	}

	if (parkInProgress && !movingRA && !movingDEC) {
		std::cout << "[PARK COMPLETE] Stopping tracking / motors\n";

		SendCommand("AA ST;");
		Sleep(100);
		SendCommand("AA ST;");

		Xstate = StateVar::Off;
		Ystate = StateVar::Off;

		CoordMem->RAGoto = 0.0;
		CoordMem->DecGoto = 0.0;

		parkInProgress = false;
		parkCompleted = true;

		std::cout << "[PARK COMPLETE] Xstate/Ystate set to Off\n";
		return;
	}

	HandleHandPaddle();
}

void PrintCoordState(Coord* c) {
	static Coord last = {};
	static bool first = true;

	if (first ||
		c->RA != last.RA ||
		c->Dec != last.Dec ||
		c->RAGoto != last.RAGoto ||
		c->DecGoto != last.DecGoto ||
		c->RASync != last.RASync ||
		c->DecSync != last.DecSync) {

		std::cout
			<< "RA=" << c->RA
			<< " Dec=" << c->Dec
			<< " RAGoto=" << c->RAGoto
			<< " DecGoto=" << c->DecGoto
			<< " RASync=" << c->RASync
			<< " DecSync=" << c->DecSync
			<< std::endl;

		last = *c;
		first = false;
	}
}

void AddTicks() {
	std::cout << "Quadrant: " << quadrant << "\n";

	if (yPole > 0) {
		std::cout << "Y Polarity: +\n";
	}
	else {
		std::cout << "Y Polarity: -\n";
	}
}

void Paddle_Timer_Update() {
	HandleHandPaddle();
}